### PR TITLE
Keep last completed segments for retention

### PR DESCRIPTION
**Pull request by jtao15**
Do not delete last completed segments in retention manager such that there's no issue to create new consuming segments if the table is paused and resumed in following cases:
1. The last completed segment is processed by MergeRollupTask, retention manager removes the last completed segment but not the 2nd last one. 
2. In rare case, the 2nd last completed segment has larger time value the the last one. The retention manager can remove the last completed segment before the 2nd last one. 
The table will start from an old offset if the consumption is resumed, and have duplicates.

Tested locally and verified the last completed segments are not removed.

### CODE REVIEW

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/retention/RetentionManagerTest.java [hunk 1]
```diff
@@ -233,6 +233,39 @@ public void testRealtimeLLCCleanup()
...
+    retentionManager.start();
+    retentionManager.run();
+
+    SegmentDeletionManager deletionManager = pinotHelixResourceManager.getSegmentDeletionManager();
+
+    // Verify that the removeAgedDeletedSegments() method in deletion manager is actually called.
+    verify(deletionManager, times(1)).removeAgedDeletedSegments(leadControllerManager);
+
+    // Verify that the deleteSegments method is actually called.
+    verify(pinotHelixResourceManager, times(1)).deleteSegments(anyString(), anyList());
```
Comment (sajjad-moradi) [inline pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/retention/RetentionManagerTest.java:266]:
```suggestion
    String tableNameWithType = tableConfig.getTableName();
    String deletedSegmentName = new LLCSegmentName("testTable", 5, 0, now).getSegmentName();
    verify(pinotHelixResourceManager, times(1)).deleteSegments(tableNameWithType, ImmutableList.of(deletedSegmentName));
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 2]
```diff
@@ -872,6 +874,24 @@ public List<SegmentZKMetadata> getSegmentsZKMetadata(String tableNameWithType) {
     return ZKMetadataProvider.getSegmentsZKMetadata(_propertyStore, tableNameWithType);
   }
 
+  public List<String> getLastLLCCompletedSegments(String tableNameWithType) {
+    Map<Integer, String> partitionIdToLastLLCCompletedSegmentMap = new HashMap<>();
+    List<SegmentZKMetadata> realtimeSegments = getSegmentsZKMetadata(tableNameWithType);
+    for (SegmentZKMetadata segMetadata : realtimeSegments) {
```
Comment (sajjad-moradi) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:880]:
```suggestion
    for (SegmentZKMetadata segMetadata : getSegmentsZKMetadata(tableNameWithType)) {
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 3]
```diff
@@ -872,6 +874,24 @@ public List<SegmentZKMetadata> getSegmentsZKMetadata(String tableNameWithType) {
...
+      if (SegmentName.isLowLevelConsumerSegmentName(segMetadata.getSegmentName())
+          && segMetadata.getStatus() == CommonConstants.Segment.Realtime.Status.DONE) {
+        LLCSegmentName llcName = LLCSegmentName.of(segMetadata.getSegmentName());
+        int partitionGroupId = llcName.getPartitionGroupId();
+        int sequenceNumber = llcName.getSequenceNumber();
+        String curSegName = partitionIdToLastLLCCompletedSegmentMap.get(partitionGroupId);
```
Comment (sajjad-moradi) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:886]: Maybe `s/curSegName/lastCompletedSegName/`

**Review by sajjad-moradi — APPROVED** [review]
LGTM, with a few minor suggestions.

**Review by sajjad-moradi — APPROVED** [review]
LGTM, with a few minor suggestions.

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/retention/RetentionManagerTest.java (hunk 1 above)
Comment (jtao15) [reply pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/retention/RetentionManagerTest.java:266]: The deleted segments are checked in the setupPinotHelixResourceManager() as following, so I think this is not needed.
```
// If and when PinotHelixResourceManager.deleteSegments() is invoked, make sure that the segments deleted
    // are exactly the same as the ones we expect to be deleted.
    doAnswer(new Answer() {
      @Override
      public Object answer(InvocationOnMock invocationOnMock)
          throws Throwable {
        Object[] args = invocationOnMock.getArguments();
        String tableNameArg = (String) args[0];
        Assert.assertEquals(tableNameArg, tableNameWithType);
        List<String> segmentListArg = (List<String>) args[1];
        Assert.assertEquals(segmentListArg.size(), removedSegments.size());
        for (String segmentName : removedSegments) {
          Assert.assertTrue(segmentListArg.contains(segmentName));
        }
        return null;
      }
    }).when(resourceManager).deleteSegments(anyString(), ArgumentMatchers.anyList());
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 4]
```diff
@@ -872,6 +874,24 @@ public List<SegmentZKMetadata> getSegmentsZKMetadata(String tableNameWithType) {
...
+        int partitionGroupId = llcName.getPartitionGroupId();
+        int sequenceNumber = llcName.getSequenceNumber();
+        String lastCompletedSegName = partitionIdToLastLLCCompletedSegmentMap.get(partitionGroupId);
+        if (lastCompletedSegName == null
+            || LLCSegmentName.of(lastCompletedSegName).getSequenceNumber() < sequenceNumber) {
+          partitionIdToLastLLCCompletedSegmentMap.put(partitionGroupId, segMetadata.getSegmentName());
+        }
+      }
+    }
+    return new ArrayList<>(partitionIdToLastLLCCompletedSegmentMap.values());
```
Comment (jackjlli) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:892]: you can just return `partitionIdToLastLLCCompletedSegmentMap.values()`, no need to wrap with another array list.

**Review by jackjlli — COMMENTED** [review]
Could you elaborate a bit on the issue we faced in the PR description section?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java (hunk 4 above)
Comment (jtao15) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java]: Removed the ArrayList and updated the return type as well.

**Comment by jtao15** [comment]
> Could you elaborate a bit on the issue we faced in the PR description section?

Updated the description to make it clearer. Please check and see if it makes sense to you.

**Review by jackjlli — APPROVED** [review]

**Review by sajjad-moradi — APPROVED** [review]

**Pull request merged**